### PR TITLE
Video Optimization: First Time Usage Dialog

### LINKED ISSUE #6087: Video Optimization: First Time Usage Dialog
## Feature Description

By default, video optimization is set to **opt in**. When the user encounters this feature for the first time, they will see a dialog with an explanation.

The dialog is displayed while optimization is already in prgoress, so it does not pause/stop the process or anything.

**Note:** "For the first time" means this is recorded in localStorage, not user meta.

Dialog text:

> Video optimization in progress
>
> Optimization is automatically enabled for all videos to ensure smooth playback. You can disable optimization any time in Settings. 
> We automatically optimize videos for smooth playback. To opt out, go to Settings.
>
> [Disable optimization] [Sounds good]

* [Disable optimization] - this will directly update the video optimization opt-in user meta value
* [Sounds good] - this will simply close the dialog

**Note:** exact wording TBD.

Screenshot 2021-02-15 at 18 16 02

**[View on Figma](https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories-Post-Stable?node-id=5511%3A441091)**

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

* If the user is encountering video optimization for the first time, display snackbar with special wording
* Telemetry: click on settings page link is being tracked

## Implementation Brief

**`useTranscodeVideo.js`**

Export `isProcessing` state (see #6066)

**`useUploadMedia.js`**

Don't show snackbar message if it's the first time usage (see #6066)

**`mediaPane.js`**

This seems like a good fit to display a new `VideoOptimizationInfoDialog`.

* Consume `isProcessing` state from `useTranscodeVideo()`
* Check `localStore.getItemByKey()` to see if dialog was already shown
* If dialog was never shown, and `isProcessing` is true, display dialog

Something like this:

```js
<VideoOptimizationInfoDialog
  open={isProcessing & isFirstTimeUsage}
  onClose={updateLocalStore}
/>
```

**`VideoOptimizationInfoDialog`**

Could live as `videoOptimizationDialog.js` next to `mediaPane.js`.

Renders a dialog with two actions: one to acknowledge info and simply close dialog (`onClose`), the other one to opt out of video processing. For the  latter, use the `useCurrentUser` hook:

```js
const { updateCurrentUser } = useCurrentUser(
  ({ actions }) => ({
    updateCurrentUser: actions.updateCurrentUser,
  })
);

// ...

updateCurrentUser({
  meta: {
    web_stories_media_optimization: false,
  },
});
```

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

By default, video optimization is set to opt in. When the user encounters this feature for the first time, they will see a dialog with an explanation.

The dialog is displayed while optimization is already in prgoress, so it does not pause/stop the process or anything.

Note: "For the first time" means this is recorded in localStorage, not user meta.

Required https://github.com/google/web-stories-wp/pull/6429 to be merged first. 

## Relevant Technical Choices

Exposing `isTranscoding` in media provider, forces the re-render that is required. 
Also check if `isTranscodingEnabled` in user meta as well, just incase the user has opted out in settings. 

## User-facing changes

Screenshot 2021-02-19 at 17 49 58

### QA

This PR can be tested by following these steps:

1. Enable video transcoding experiment
2. Open editor. 
3. Drag a video file into editor. 
4. See dialog.  

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6087

### CODE REVIEW

**Comment by spacedmonkey** [comment]
Should I add a storybook intrgration?

**Comment by swissspidy** [comment]
> Should I add a storybook intrgration?

Yes, all dialogs should be in the storybook

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+  const dialogTitle = __('Video optimization in progress', 'web-stories');
+  const dialogDescription = __(
+    'This process can’t be stopped once it begins. Optimization is automatically enabled for all videos to ensure smooth playback. You can disable optimization any time in Settings.',
+    'web-stories'
+  );
+
+  return (
+    <Dialog
+      open={isTranscodingEnabled && isTranscoding && !hasOptedIn}
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: The `isTranscodingEnabled` check is redundant.

`isTranscoding` can only ever be `true` if `isTranscodingEnabled` is also `true`.

```suggestion
      open={isTranscoding && !hasOptedIn}
```

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    isTranscoding: state.state.isTranscoding,
+  }));
+
+  const [hasOptedIn, setHasOptedIn] = useState(
+    Boolean(
+      localStore.getItemByKey(
+        `${LOCAL_STORAGE_PREFIX.VIDEO_OPTIMIZATION_OPTIN}`
+      )
+    )
+  );
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: I'm not super fond of the `optin`/`optedIn` naming here.

This doesn't really track whether the user has _opted in to video optimization_, but whether the dialog has been displayed already.

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: This is from an earlier version of this PR, sorry forgot to remove it - https://github.com/google/web-stories-wp/pull/6443/commits/fd50047cb2524ddc25e84d6d57435fa38d83de47

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 2 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: Changed variable name to this. Let me know if you are happy with this new name. https://github.com/google/web-stories-wp/pull/6443/commits/2033a19a26fe3cdc025208d4869199715683d0db

**Comment by spacedmonkey** [comment]
I have changed variable names and added story book integrations. 

The story book integration doesn't work very well, as this component relays on localStorage. I really think this is a good reason to save this setting in user meta, it makes it easier to mock and the code much cleaner here.

File: assets/src/edit-story/components/library/panes/media/local/test/videoOptimizationDialog.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+describe('videoOptimizationDialog', () => {
+  it('should render', () => {
+    const { queryByText, getByRole } = setup();
+
+    expect(queryByText('Video optimization in progress')).toBeInTheDocument();
+    expect(getByRole('button', { name: /Disable/i })).toBeInTheDocument();
+    expect(getByRole('button', { name: /Sounds/i })).toBeInTheDocument();
+  });
+  it('should clicking disable will trigger api request to update user meta', async () => {
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/test/videoOptimizationDialog.js]:
```suggestion
  it('should trigger API request to update user when disabling', async () => {
```

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import { useLocalMedia } from '../../../../../app/media';
+
+function VideoOptimizationDialog() {
+  const { isTranscoding } = useLocalMedia((state) => ({
+    isTranscoding: state.state.isTranscoding,
+  }));
+
+  const KEY = LOCAL_STORAGE_PREFIX.VIDEO_OPTIMIZATION_DIALOG_DISMISSED;
+
+  const [dialogDismissed, setDialogDismissed] = useState(
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]:
```suggestion
  const [isDialogDismissed, setIsDialogDismissed] = useState(
```

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+} from '../../../../../utils/localStore';
+import { useCurrentUser } from '../../../../../app/currentUser';
+import { useLocalMedia } from '../../../../../app/media';
+
+function VideoOptimizationDialog() {
+  const { isTranscoding } = useLocalMedia((state) => ({
+    isTranscoding: state.state.isTranscoding,
+  }));
+
+  const KEY = LOCAL_STORAGE_PREFIX.VIDEO_OPTIMIZATION_DIALOG_DISMISSED;
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: No need to uppercase this const.

```suggestion
  const storageKey = LOCAL_STORAGE_PREFIX.VIDEO_OPTIMIZATION_DIALOG_DISMISSED;
```

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  const [dialogDismissed, setDialogDismissed] = useState(
+    Boolean(localStore.getItemByKey(`${KEY}`))
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]:
```suggestion
    Boolean(localStore.getItemByKey(storageKey))
```

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 5 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: Also, could move it outside the function. Then you don't need to add it as dependency to the `useCallback`

**Comment by swissspidy** [comment]
> I have changed variable names and added story book integrations.
> 
> The story book integration doesn't work very well, as this component relays on localStorage. I really think this is a good reason to save this setting in user meta, it makes it easier to mock and the code much cleaner here.

Not a good enough reason to change a product requirement :-)

**Comment by spacedmonkey** [comment]
> Not a good enough reason to change a product requirement :-)

I disagreed with the product requirements before. But I don't care enough to fight the point here.

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 6 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: Fixed in https://github.com/google/web-stories-wp/pull/6443/commits/1fdf2e87c86fd1ab2367b22af4acb122853f1152

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 5 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: Fixed in https://github.com/google/web-stories-wp/pull/6443/commits/1fdf2e87c86fd1ab2367b22af4acb122853f1152

File: assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/local/videoOptimizationDialog.js]: Fixed https://github.com/google/web-stories-wp/pull/6443/commits/1fdf2e87c86fd1ab2367b22af4acb122853f1152

**Review by swissspidy — APPROVED** [review]

**Pull request merged**